v1.10 backports 2021-09-14

kaworu opened the pull request:
* #17275 -- istio: Update to release 1.10.4 (@jrajahalme) :heavy_check_mark:
 * #17445 (@jrajahalme) :heavy_check_mark:
 * #17338 -- fix(docs): bandwidth-manager install error (@withlin) :heavy_check_mark: 
 * #17351 -- test/runtime: Look into log errors after test start (@joamaki) :heavy_check_mark: 
 * #17168 -- datapath: Do not SNAT replies to outside (@brb) :heavy_check_mark: 
 * #17153 -- ethtool: use ioctl wrapper from golang.org/x/sys/unix (@tklauser) :heavy_check_mark: 
 * #17381 -- docs: Clarify exact requirements for the egress gateway (@pchaigno) :heavy_check_mark: 
 * #17258 -- jenkinsfiles: Don't display nulls in current build display name (@twpayne) :heavy_check_mark: 
 * #17143 (@nbusseneau) :heavy_check_mark:
 * #17464 (@nebril)

Once this PR is merged, you can update the PR labels via:
```upstream-prs
$ for pr in 17275 17445 17338 17351 17168 17153 17381 17258 17143 17464; do contrib/backporting/set-labels.py $pr done 1.10; done
```

jrajahalme committed:
istio: Update to release 1.10.4

[ upstream commit 4c87394b7a2380909d86ffa3e8da47df0fff2e98 ]

Update Cilium Istio integration to Istio release 1.10.4.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>
Signed-off-by: Alexandre Perrin <alex@kaworu.ch>

nbusseneau committed:
workflows: pin `cilium-cli` version to v0.8.6

[ upstream commit 2202dae9a4e74fc1d485d5da48484cb7ae79671b ]

In #16892, we switched from pinning CLI version in workflows to using
the latest stable version automatically. This can cause issues if a new
release does not play nice with the set of environments tested by the
workflows on `cilium/cilium`.

We are reverting to pinning CLI version so as to have better control
over the test environment, and avoid new CLI releases negatively
impacting `cilium/cilium` workflows immediately upon release.

With the CLI version pinned, any issues with the new version will be
detected in the PR bumping the pinned version, allowing us to fix them
prior to merging.

Signed-off-by: Nicolas Busseneau <nicolas@isovalent.com>
Signed-off-by: Alexandre Perrin <alex@kaworu.ch>

brb committed:
datapath: Do not SNAT replies to outside

[ upstream commit d3ff998f2b30b41227a395b6415627d8cb611e69 ]

Previously, the BPF-based masquerading (--enable-bpf-masquerade=true)
was wrongly masquerading replies from a pod to an outside when the
outside had initiated a connection. This was possible when e.g., the
outside had a route to the pod cidr.

To fix this, we introduce a lightweight CT lookup function
ct_is_reply4() which checks whether a given flow is a reply. The lookup
function is called in snat_v4_needed().

As a side note, I've tried to move the port extraction to a separate
function, but unfortunately it hits complexity issues on the 4.19
kernel in the "K8sDatapathConfig AutoDirectNodeRoutes Check direct
connectivity with per endpoint routes" suite:

    BPF program is too large. Processed 131073 insn
    libbpf: failed to load program 'handle_to_container'
    libbpf: failed to load object '624_next/bpf_lxc.o'

Signed-off-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Alexandre Perrin <alex@kaworu.ch>

brb committed:
datapath: Improve snat_v4_needed() comment

[ upstream commit 55bfba9dd5b574bacb5b4afbfa337fc9d4cdd5cf ]

Signed-off-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Alexandre Perrin <alex@kaworu.ch>

brb committed:
test: Enable BPF masq in test from k8s

[ upstream commit 4b92d2de7f36de246dd5b2144f1e8d2532f23f0e ]

Previously, they were failing due to our datapath masquerading replies
from pod to outside. As it got fixed in the previous commit, we can
enable BPF-based masquerading.

This will also gives us some coverage for the fix.

Signed-off-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Alexandre Perrin <alex@kaworu.ch>

tklauser committed:
vendor: bump golang.org/x/sys to latest version

[ upstream commit 6418ade4ba021ca5d8b46dac70d707617e5dd0a4 ]

This pulls in a few fixes around ioctl wrappers wrt. unsafe.Pointer
usage and fixes ifreqEthtool to be correctly padded.

Ref. golang/sys@e5e7981a10699f0af2ffea4c0e0f542e447b2e4a
Ref. golang/sys@b4502255bfe74b77ffdeb6ed972eeecd98d13721

Signed-off-by: Tobias Klauser <tobias@cilium.io>
Signed-off-by: Alexandre Perrin <alex@kaworu.ch>

tklauser committed:
datapath/linux/ethtool: use IoctlGetEthtoolDrvinfo

[ upstream commit d82ac6f54c0118088cc46d8d892ff5e87cf5d09e ]

Use the ioctl wrapper provided in the golang.org/x/sys/unix package with
the correctly padded ifreqData struct, rather than providing our own
wrapper and struct which is incorrectly padded.

Also add a simple unit test and make sure the package is only built on
Linux.

Signed-off-by: Tobias Klauser <tobias@cilium.io>
Signed-off-by: Alexandre Perrin <alex@kaworu.ch>

joamaki committed:
test/runtime: Look into log errors after test start

[ upstream commit 38994b01cdd6b707a422263e379692dcacaf14e1 ]

When running runtime tests locally sometimes the test fail as level=error
log entries are found that are the result of cilium-agent restarts during
provisioning.

This is similar to the fix done in https://github.com/cilium/cilium/pull/14529.

Signed-off-by: Jussi Maki <jussi@isovalent.com>
Signed-off-by: Alexandre Perrin <alex@kaworu.ch>

withlin committed:
fix bandwidth-manager install error Signed-off-by: JinLin Fu withlin@apache.org

[ upstream commit 712af8e26e4afe1f380ee081b152fba1392b3b96 ]

Signed-off-by: Alexandre Perrin <alex@kaworu.ch>

twpayne committed:
jenkinsfiles: Don't display nulls in current build display name

[ upstream commit e0da2e441e506279bdc9d3089024655abfbebb8d ]

Signed-off-by: Tom Payne <tom@isovalent.com>
Co-authored-by: Nicolas Busseneau <nicolas@isovalent.com>
Signed-off-by: Alexandre Perrin <alex@kaworu.ch>

pchaigno committed:
docs: Clarify exact requirements for the egress gateway

[ upstream commit 82469c3f79ca326f8dbae2e7374a8a9fdc99a125 ]

The egress gateway doesn't exactly require our kube-proxy replacement to
be enabled. It only requires BPF masquerading which itself requires BPF
NodePort. Enabling KPR is just an easy way to enable BPF NodePort.

Signed-off-by: Paul Chaignon <paul@cilium.io>
Signed-off-by: Alexandre Perrin <alex@kaworu.ch>

tklauser commented on test/k8sT/istio.go:
```diff
@@ -38,7 +38,11 @@ var _ = SkipDescribeIf(helpers.RunsOn54Kernel, "K8sIstioTest", func() {
 		istioVersion = "1.8.2"
+=======
+		istioVersion = "1.10.4"
+>>>>>>> istio: Update to release 1.10.4
```
Looks like an unresolved merge conflict.

pchaigno replied:
(I usually run `git rebase --exec "make clean && make -j6 && make -C test -j6 && make -C bpf build_all -j6" origin/v1.10` locally to catch such issues.)

kaworu replied:
Good catch @tklauser and thanks for the command line @pchaigno!

I usually runs some checks locally, but the `//go:build` introduced by Go v1.17 make it so `contrib/scripts/check-fmt.sh` fails quickly. Setting `GO_VERSION` doesn't help here since it's `gofmt`, and the script doesn't allow to override the `gofmt` binary path (e.g. to `~/sdk/go1.16.7/bin/gofmt`). Have you a workaround for this? Should we have one?

pchaigno reviewed (approved):
> - fix(docs): bandwidth-manager install error #17338 -- fix(docs): bandwidth-manager install error (@withlin)
> - docs: Clarify exact requirements for the egress gateway #17381 -- docs: Clarify exact requirements for the egress gateway (@pchaigno)

Those two look good to me.

tklauser reviewed (approved):
> * #17153 -- ethtool: use ioctl wrapper from golang.org/x/sys/unix (@tklauser)

This one LGTM.

joamaki reviewed (approved):
> test/runtime: Look into log errors after test start #17351 

LGTM.

nbusseneau reviewed (approved):
LGTM!

twpayne reviewed (approved):
LGTM for my changes

brb reviewed (approved):
My changes LGTM, thanks!

jrajahalme committed:
test: Skip Istio test on k8s <1.17

[ upstream commit 3992048f6ac44b2e44f4c2d9d157a3987b883af0 ]

Istio 1.10 requires at least k8s version 1.17.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>
Signed-off-by: Alexandre Perrin <alex@kaworu.ch>

nebril committed:
ci: update cilium-cli to 0.9.1

[ upstream commit acf34318e49df9060fc51b205d1a0069c988ee62 ]

This change updates cilium-cli to 0.9.1 in github action workflows
files.

Signed-off-by: Maciej Kwiek <maciej@isovalent.com>
Signed-off-by: Alexandre Perrin <alex@kaworu.ch>

jrajahalme reviewed (approved):
Istio backports look OK.
